File: packages/shared/src/types/emergency.ts
/** Emergency arrival mode */
export type ArrivalMode = "walk_in" | "ambulance" | "referral" | "police" | "other";

/** SATS triage category (emergency-specific; see encounter.ts for general TriageCategory) */
export type EmergencyTriageCategory = "emergency" | "urgent" | "standard" | "non_urgent" | "dead";

/** SATS triage color */
export type TriageColor = "red" | "orange" | "yellow" | "green" | "blue";

/** Emergency treatment area */
export type TreatmentArea = "resus" | "acute" | "sub_acute" | "fast_track" | "observation" | "paediatric";

/** Emergency visit status */
export type EmergencyVisitStatus =
  | "arrived"
  | "triaged"
  | "in_treatment"
  | "observation"
  | "admitted"
  | "discharged"
  | "transferred"
  | "deceased"
  | "left_against_advice";

/** Disposition type */
export type DispositionType = "discharge" | "admit" | "transfer" | "deceased" | "left_ama";

/** Emergency visit creation request */
export interface EmergencyVisitCreate {
  patient_id: string;
  arrival_mode?: ArrivalMode;
  brought_by?: string | null;
  chief_complaint: string;
  is_trauma?: boolean;
  allergies_noted?: string | null;
  notes?: string | null;
}

/** Emergency visit response */
export interface EmergencyVisitResponse {
  id: string;
  patient_id: string;
  encounter_id: string | null;
  visit_number: string;
  arrival_time: string;
  arrival_mode: string;
  brought_by: string | null;
  chief_complaint: string;
  triage_category: string;
  triage_color: string;
  triage_score: number | null;
  triage_time: string | null;
  triage_vitals: Record<string, unknown> | null;
  assigned_doctor_id: string | null;
  treatment_area: string | null;
  treatment_started_at: string | null;
  status: string;
  disposition: string | null;
  disposition_time: string | null;
  disposition_notes: string | null;
  is_trauma: boolean;
  is_resuscitation: boolean;
  allergies_noted: string | null;
  interventions: Record<string, unknown> | null;
  notes: string | null;
  created_at: string;
}

/** Emergency queue list item */
export interface EmergencyListItem {
  id: string;
  visit_number: string;
  patient_id: string;
  patient_name: string | null;
  patient_mrn: string | null;
  arrival_time: string;
  arrival_mode: string;
  chief_complaint: string;
  triage_category: string;
  triage_color: string;
  triage_score: number | null;
  treatment_area: string | null;
  assigned_doctor_name: string | null;
  status: string;
  is_trauma: boolean;
  is_resuscitation: boolean;
}

/** Emergency queue response */
export interface EmergencyListResponse {
  items: EmergencyListItem[];
  total: number;
}

/** Triage request */
export interface TriageRequest {
  triage_category: EmergencyTriageCategory;
  triage_score?: number | null;
  triage_vitals?: Record<string, unknown> | null;
  treatment_area?: TreatmentArea | null;
  notes?: string | null;
}

/** Assign doctor request */
export interface AssignDoctorRequest {
  doctor_id: string;
}

/** Disposition request */
export interface DispositionRequest {
  disposition: DispositionType;
  disposition_notes?: string | null;
  admitted_to_ward_id?: string | null;
}

/** Emergency department summary */
export interface EmergencySummary {
  total_today: number;
  awaiting_triage: number;
  in_treatment: number;
  in_observation: number;
  critical_red: number;
  urgent_orange: number;
  discharged_today: number;
  admitted_today: number;
  trauma_cases: number;
}

File: services/api-gateway/app/schemas/emergency.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field


class EmergencyVisitCreate(BaseModel):
    """Schema for registering an emergency visit."""

    patient_id: uuid.UUID
    arrival_mode: str = Field(
        default="walk_in",
        pattern=r"^(walk_in|ambulance|referral|police|other)$",
    )
    brought_by: str | None = Field(None, max_length=200)
    chief_complaint: str = Field(..., max_length=2000)
    is_trauma: bool = False
    allergies_noted: str | None = Field(None, max_length=500)
    notes: str | None = Field(None, max_length=2000)


class TriageRequest(BaseModel):
    """Schema for triaging an emergency patient."""

    triage_category: str = Field(
        ...,
        pattern=r"^(emergency|urgent|standard|non_urgent|dead)$",
    )
    triage_score: int | None = Field(None, ge=0, le=17)
    triage_vitals: dict | None = None
    treatment_area: str | None = Field(
        None,
        pattern=r"^(resus|acute|sub_acute|fast_track|observation|paediatric)$",
    )
    notes: str | None = Field(None, max_length=1000)


class AssignDoctorRequest(BaseModel):
    """Schema for assigning a doctor to an emergency visit."""

    doctor_id: uuid.UUID


class DispositionRequest(BaseModel):
    """Schema for recording disposition."""

    disposition: str = Field(
        ...,
        pattern=r"^(discharge|admit|transfer|deceased|left_ama)$",
    )
    disposition_notes: str | None = Field(None, max_length=2000)
    admitted_to_ward_id: uuid.UUID | None = None


class EmergencyVisitResponse(BaseModel):
    """Schema for emergency visit API responses."""

    id: uuid.UUID
    patient_id: uuid.UUID
    encounter_id: uuid.UUID | None
    visit_number: str
    arrival_time: datetime
    arrival_mode: str
    brought_by: str | None
    chief_complaint: str
    triage_category: str
    triage_color: str
    triage_score: int | None
    triage_time: datetime | None
    triage_vitals: dict | None
    assigned_doctor_id: uuid.UUID | None
    treatment_area: str | None
    treatment_started_at: datetime | None
    status: str
    disposition: str | None
    disposition_time: datetime | None
    disposition_notes: str | None
    is_trauma: bool
    is_resuscitation: bool
    allergies_noted: str | None
    interventions: dict | None
    notes: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class EmergencyListItem(BaseModel):
    """Emergency visit in queue list with patient name."""

    id: uuid.UUID
    visit_number: str
    patient_id: uuid.UUID
    patient_name: str | None = None
    patient_mrn: str | None = None
    arrival_time: datetime
    arrival_mode: str
    chief_complaint: str
    triage_category: str
    triage_color: str
    triage_score: int | None
    treatment_area: str | None
    assigned_doctor_name: str | None = None
    status: str
    is_trauma: bool
    is_resuscitation: bool

    model_config = {"from_attributes": True}


class EmergencyListResponse(BaseModel):
    """Emergency queue response."""

    items: list[EmergencyListItem]
    total: int


class EmergencySummary(BaseModel):
    """Emergency department summary stats."""

    total_today: int = 0
    awaiting_triage: int = 0
    in_treatment: int = 0
    in_observation: int = 0
    critical_red: int = 0
    urgent_orange: int = 0
    discharged_today: int = 0
    admitted_today: int = 0
    trauma_cases: int = 0
